File: collect_app/src/test/java/org/odk/collect/android/support/TestActivityScenario.java
package org.odk.collect.android.support;

import android.app.Activity;

import androidx.test.core.app.ActivityScenario;

import org.robolectric.Robolectric;
import org.robolectric.android.controller.ActivityController;

/**
 * Allows the use of an {@link androidx.test.core.app.ActivityScenario} like interface when
 * using an Activity that isn't part of the manifest. This is useful when using a mock/fake/stub
 * Activity as part of a test
 */
public final class TestActivityScenario<T extends Activity> {

    private final ActivityController<T> activityController;

    public static <A extends Activity> TestActivityScenario<A> launch(Class<A> activityClass) {
        TestActivityScenario<A> activityScenario = new TestActivityScenario<>(activityClass);
        activityScenario.launchInternal();
        return activityScenario;
    }

    private TestActivityScenario(Class<T> activityClass) {
        activityController = Robolectric.buildActivity(activityClass);
    }

    private void launchInternal() {
        activityController.setup();
    }

    public void onActivity(ActivityScenario.ActivityAction<T> activityAction) {
        activityAction.perform(activityController.get());
    }

    public void recreate() {
        activityController.recreate();
    }
}

File: collect_app/src/test/java/org/odk/collect/android/support/TestActivityScenario.kt
package org.odk.collect.android.support

import android.app.Activity
import androidx.test.core.app.ActivityScenario
import org.robolectric.Robolectric
import org.robolectric.android.controller.ActivityController

/**
 * Allows the use of an [androidx.test.core.app.ActivityScenario] like interface when
 * using an Activity that isn't part of the manifest. This is useful when using a mock/fake/stub
 * Activity as part of a test
 */
class TestActivityScenario<T : Activity?> private constructor(activityClass: Class<T>) {
    private val activityController: ActivityController<T> = Robolectric.buildActivity(activityClass)

    private fun launchInternal() {
        activityController.setup()
    }

    fun onActivity(activityAction: ActivityScenario.ActivityAction<T>) {
        activityAction.perform(activityController.get())
    }

    fun recreate() {
        activityController.recreate()
    }

    companion object {
        @JvmStatic
        fun <A : Activity?> launch(activityClass: Class<A>): TestActivityScenario<A> {
            val activityScenario = TestActivityScenario(activityClass)
            activityScenario.launchInternal()
            return activityScenario
        }
    }
}
